File: golang/leetcode/strStr.go
package leetcode

// Implement strStr().
// Returns the index of the first occurrence of needle in haystack, or -1 if needle is not part of haystack.
//
// Subscribe to see which companies asked this question
//
// Tags
// Two Pointers String
//
// Related Topics
// Two Pointers

func strStr(haystack string, needle string) int {
	if len(needle) == 0 {
		return 0
	}
	if len(haystack) == 0 {
		return -1
	}

	for i := 0; ; i++ {
		for j := 0; ; j++ {
			if j == len(needle) {
				return i
			}
			if i+j == len(haystack) {
				return -1
			}
			if haystack[i+j] != needle[j] {
				break
			}
		}
	}
}

// test
func TestStrStr(t *testing.T) {
	cases := []struct {
		haystack string
		needle   string
		expect   int
	}{
		{"hello", "ll", 2},
		{"aaaaa", "bba", -1},
		{"aaaaa", "", 0},
		{"", "", 0},
		{"", "a", -1},
	}

	for _, c := range cases {
		if actual := strStr(c.haystack, c.needle); actual != c.expect {
			t.Fatalf("%s, %s, expect %d, actual %d\n", c.haystack, c.needle, c.expect, actual)
		}
	}
}

File: golang/leetcode/rust/strStr.rs
fn strStr(haystack: String, needle: String) -> i32 {
    let haystack_bytes = haystack.as_bytes();
    let needle_bytes = needle.as_bytes();
    let mut i = 0;
    let mut j = 0;
    while i < haystack_bytes.len() && j < needle_bytes.len() {
        if haystack_bytes[i] != needle_bytes[j] {
            i += 1;
        } else {
            i += 1;
            j += 1;
        }
    }
    if j == needle_bytes.len() {
        return i - j;
    } else {
        return -1;
    }
}

// test
fn main() {
    assert_eq!(strStr("hello".to_string(), "ll".to_string()), 2);
    assert_eq!(strStr("aaaaa".to_string(), "bba".to_string()), -1);
    assert_eq!(strStr("aaaaa".to_string(), "aaaaa".to_string()), 0);
    assert_eq!(strStr("".to_string(), "".to_string()), 0);
    assert_eq!(strStr("".to_string(), "a".to_string()), -1);
    assert_eq!(strStr("a".to_string(), "".to_string()), 0);
    assert_eq!(strStr("a".to_string(), "a".to_string()), 0);
    assert_eq!(strStr("a".to_string(), "b".to_string()), -1);
    assert_eq!(strStr("b".to_string(), "a".to_string()), -1);
    assert_eq!(strStr("b".to_string(), "b".to_string()), 0);
    assert_eq!(strStr("b".to_string(), "c".to_string()), -1);
    assert_eq!(strStr("c".to_string(), "b".to_string()), -1);
    assert_eq!(strStr("c".to_string(), "c".to_string()), 0);
    assert_eq!(strStr("c".to_string(), "d".to_string()), -1);
    assert_eq!(strStr("d".to_string(), "c".to_string()), -1);
    assert_eq!(strStr("d".to_string(), "d".to_string()), 0);
    assert_eq!(strStr("d".to_string(), "e".to_string()), -1);
    assert_eq!(strStr("e".to_string(), "d".to_string()), -1);
    assert_eq!(strStr("e".to_string(), "e".to_string()), 0);
}